process_metrics_linux.go

```diff
@@ -89,8 +95,34 @@ func writeIOMetrics(w io.Writer) {
 	ioFilepath := "/proc/self/io"
 	data, err := ioutil.ReadFile(ioFilepath)
 	if err != nil {
-		log.Printf("ERROR: metrics: cannot open %q: %s", ioFilepath, err)
+		// Do not spam the logs with errors
+		// This error will not be fixed without process restart
+		var reportErr func()
+		if os.IsNotExist(err) {
```

Comment:
I suggest to use switch case.

 And current implementation never resolve metrics state change, because it happens with sync.Once.


```go
// use default values
 var processSelfIONotFoundErrorCount,processSelfIOPermErrorCount,processSelfErrorCount uint64

 if err != nil {
 var errMsg string
switch {
   case os.IsNotExist(err):
         errMsg = fmt.Sprintf("ERROR: metrics: cannot open %q: %s. This is expected on kernel without CONFIG_TASK_IO_ACCOUNTING, systems without cgroup controller for IO.")
          processSelfIONotFoundErrorCount = 1
  case os.IsPermission(err):
        errMsg = ...
        processSelfIOPermErrorCount =  1
  default:
         errMsg  = ..
         processSelfErrorCount = 1
}
reportIoError.Do(func(){ log.Println(errMsg) )
}

// always export  metrics even if err == nil. It's good practice for prometheus metrics
// Use gauge instead of counter. It indicates current state, I'm not sure that we need rate for this kind of errors.
fmt.Fprintf(w, "process_io_stats_read_error{reason=\"not_found\"} %d\n", processSelfIONotFoundErrorCount)
```